messaging.wit.md

```diff
@@ -0,0 +1,118 @@
+		id: string,
+		data: option<list<u8>>,
+		datacontenttype: option<string>,
+		dataschema: option<string>,
+		subject: option<string>,
+		time: option<string>,	
+		extensions: option<list<tuple<string, string>>>
+    }
+    
+    /// Channels specify where a published message should land. There are two types of channels:
+    /// - queue: competitive consumers, and
+    /// - topic: non-competitive consumers.
+    variant channel {
+		queue(string),
+		topic(string)
+    }
+    
+    /// An error type
+    variant messaging-error {
+        payload-too-large(string),
+        queue-or-topic-not-found(string),
+        insufficient-permissions(string),
+        service-unavailable(string),
+        delivery-failed(string),
+        connection-lost(string),
+        unsupported-message-format(string),
+        unexpected-error(string),
+    }
+
+	// A subscription token that allows receives from a specific subscription
+	type subscription-token = string
+}
+
+/// An interface for a producer.
+interface "wasi:messaging/pub" { 
+	/// Publishes an event to a channel.
+	publish: func(c: channel, e: event) -> result<_, error>
+}
+
+/// An interface for a generic consumer.
+interface "wasi:messaging/sub" {
+	/// Subscribes to a channel.
+  	subscribe: func(c: channel) -> result<subscription-token, error>
+
+	/// Unsubscribes from a channel.
+  	unsubscribe: func(st: subscription-token) -> result<_, error>
+}
+
+/// An interface for a consumer relying on push-based message delivery.
+interface "wasi:messaging/push" {
+	/// Creates an on-receive handler push-based message delivery.
+  	on-receive: func(e: event) -> result<_, error>
+}
+
+/// An interface for a consumer relying on basic pull-based message delivery.
+interface "wasi:messaging/basic/pull" {
+	/// Pulls a message from a specific channel.
+  	receive: func(time-to-live-in-milliseconds: u32, st: subscription-token) -> result<event, error>
+}
+
+/// An interface for a consumer relying on pull-based message delivery via streaming.
+interface "wasi:messaging/stream/pull" {
+	/// Pulls a stream of messages from a specific channel.
+  	stream-receive: func(st: subscription-token) -> result<stream<event>, error> 
+}
+
+// ...
+```
+
+## World Examples
+
+```wit
+// ...
+
+/// Two worlds that connect "sub" together "basic/pull"/"stream/pull", as they are intrinsically linked.
+world "wasi:messaging/basic/sub-pull" {
+	import sub: "wasi:messaging/basic/sub"
+	
+	export pull: "wasi:messaging/basic/pull"
+}
+
+world "wasi:messaging/stream/sub-pull" {
+	import sub: "wasi:messaging/basic/sub"
+	
+	export stream-pull: "wasi:messaging/stream/pull"
+}
+
+/// Two typical usage examples.
+world "wasi:messaging/pull-pubsub" {
+    import pub: "wasi:messaging/pub"
+    import sub-pull: "wasi:messaging/basic/sub-pull" // or "wasi:messaging/stream/sub-pull"
+  
+    export handle-receive: "wasi:http/handler"
+    export handle-subscribe: "wasi:http/handler"
+    export handle-publish: "wasi:http/handler"
```

Comment:
I don't think "wasi:http/handler" exposes these functions. Do you mean "export handle-http: 'wasi:http/handler'"?

Reply:
We don't really have a "wasi:http/handler" defined, but — yes, you're right. My idea here is to convery exporting "named" handlers.